### PR TITLE
Add Unit tests for authcontext

### PR DESCRIPTION
**Pull request by ChathurangiShyalika**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedMap;
+
+import static org.mockito.Matchers.anyInt;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.powermock.api.mockito.PowerMockito.*;
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Remove * imports

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+import static org.testng.Assert.*;
+
+@PrepareForTest({OAuthServerConfiguration.class, OAuth2Util.class, MultitenantUtils.class, OAuthComponentServiceHolder.class, UserRealm.class})
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Break this into two lines if it exceeds 120 characters

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        claim4.setClaimUri(userClaimUri2);
+        testMappedAttributesUser2 = "MappedAttributesUser2";
+
+        ClaimMapping cMap1 = new ClaimMapping(claim1, testMappedAttributesCore1);
+        ClaimMapping cMap2 = new ClaimMapping(claim2, testMappedAttributesCore2);
+        ClaimMapping cMap3 = new ClaimMapping(claim3, testMappedAttributesUser1);
+        ClaimMapping cMap4 = new ClaimMapping(claim4, testMappedAttributesUser2);
+
+        ClaimMapping[] coreClaims = new ClaimMapping[]{cMap1, cMap2, cMap3, cMap4};
+
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Too many new lines in between.

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    @Test
+    public void testInit() throws Exception {
+        // Subject is null.
+        assertFalse(Boolean.parseBoolean(defaultClaimsRetriever.DEFAULT_DIALECT_URI));
+
+        // Subject is not null.
+        mockStatic(OAuthServerConfiguration.class);
+        when(OAuthServerConfiguration.getInstance()).thenReturn(mockedOAuthServerConfiguration);
+        when(mockedOAuthServerConfiguration.getConsumerDialectURI()).thenReturn("consumerDialectURI");
+        assertFalse(Boolean.parseBoolean(mockedOAuthServerConfiguration.getConsumerDialectURI()));
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Is this assert correct?
It checks the value we mock and not anything from the class being unit tested

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        mockStatic(MultitenantUtils.class);
+        mockStatic(OAuthComponentServiceHolder.class);
+
+        when(OAuthComponentServiceHolder.getInstance()).thenReturn(mockedOAuthComponentServiceHolder);
+        when(mockedOAuthComponentServiceHolder.getRealmService()).thenReturn(mockedRealmService);
+        mockedUserRealm = mock(UserRealm.class);
+        when(mockedRealmService.getTenantUserRealm(anyInt())).thenReturn(mockedUserRealm);
+        when(mockedUserRealm.getUserStoreManager()).thenThrow(new UserStoreException("UserStoreException"));
+        Map<String, String> expectedMappingTrue = new HashMap<>();
+        when(mockedUserStoreManager.getUserClaimValue(anyString(), anyString(), anyString())).thenThrow(new UserStoreException("UserStoreException"));
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: break into two lines if this exceeds 120 characters.

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        // Subject is not null.
+        mockStatic(OAuthServerConfiguration.class);
+        when(OAuthServerConfiguration.getInstance()).thenReturn(mockedOAuthServerConfiguration);
+        when(mockedOAuthServerConfiguration.getConsumerDialectURI()).thenReturn("consumerDialectURI");
+        assertFalse(Boolean.parseBoolean(mockedOAuthServerConfiguration.getConsumerDialectURI()));
+    }
+
+    @Test
+    public void testGetClaims() throws Exception {
+        SortedMap<String, String> claimValues;
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: We haven't used this variable anywhere within this method. If it is not used remove it.

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        String[] claims = {"https://www.wso2.org/address", "https://www.wso2.org/email"};
+
+        assertNotNull(defaultClaimsRetriever.getClaims("admin", claims));
+    }
+
+    @Test(expectedExceptions = IdentityOAuth2Exception.class)
+    public void testGetClaims1() throws Exception {
+        SortedMap<String, String> claimValues;
+        int tenantId = MultitenantConstants.SUPER_TENANT_ID;
+
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: We haven't used this variable anywhere within this method. If it is not used remove it.

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        Map<String, String> expectedMappingTrue = new HashMap<>();
+        when(mockedUserStoreManager.getUserClaimValue("user", "[\"https://www.wso2.org/address\", \"https://www.wso2.org/email\"]", "read")).thenReturn(expectedMappingTrue.toString());
+
+        String[] claims = {"https://www.wso2.org/address", "https://www.wso2.org/email"};
+
+        assertNotNull(defaultClaimsRetriever.getClaims("admin", claims));
+    }
+
+    @Test(expectedExceptions = IdentityOAuth2Exception.class)
+    public void testGetClaims1() throws Exception {
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: would be better to give a meaningful name to the method like testGetClaimsWhenBlahBlah()

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+    public void testGetDefaultClaims() throws Exception {
+
+        mockStatic(OAuthServerConfiguration.class);
+        when(OAuthServerConfiguration.getInstance()).thenReturn(mockedOAuthServerConfiguration);
+        when(mockedOAuthServerConfiguration.getTimeStampSkewInSeconds()).thenReturn((long) 7);
+        mockStatic(OAuth2Util.class);
+        mockStatic(MultitenantUtils.class);
+        mockStatic(OAuthComponentServiceHolder.class);
+
+        when(OAuthComponentServiceHolder.getInstance()).thenReturn(mockedOAuthComponentServiceHolder);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Why did you have to mock realm service?
Is it used internally by the method tested?

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/bean/ScopeTest.java [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+public class ScopeTest {
+    Scope scope;
+    String name = "readClaims";
+    String description = "Test cases to test scope";
+
+    ArrayList<String> bindings = new ArrayList<>(Arrays.asList("scope1", "scope2"));
+
+    @BeforeTest
+    public void setUp() throws IllegalAccessException, InstantiationException {
+        scope = new Scope(name, description);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/bean/ScopeTest.java]: scope variable is assigned twice meaning the first assignment is useless.

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/bean/ScopeTest.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+        scope.addBindings(Arrays.asList(new String[]{"scope3", "scope4"}));
+        assertTrue(scope.getBindings().containsAll(Arrays.asList(new String[]{"scope3", "scope4"})));
+    }
+
+    @Test
+    public void testGetBindings() throws Exception {
+        assertNotNull(bindings);
+        bindings.add("scope4");
+        assertEquals(scope.getBindings(), bindings, "Actual size is not match with expected one");
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/bean/ScopeTest.java]: Actual size is not match with expected one --> Actual size does not match with expected one

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/bean/ScopeTest.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    @Test
+    public void testSetBindings() throws Exception {
+        ArrayList<String> bindings2 = new ArrayList<>(Arrays.asList("scope1", "scope2", "scope3"));
+        scope.setBindings(bindings2);
+        int scopeSize = 3;
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/bean/ScopeTest.java]: do we need this variable?
if want to keep it then rename it to a meaningful one like expectedScopeSize

Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/bean/ScopeTest.java]: Renamed it as expectedScopeSize

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/bean/ScopeTest.java (hunk 11 above)
Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/bean/ScopeTest.java]: Fixed

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java (hunk 5 above)
Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Done

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java (hunk 7 above)
Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Done

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java (hunk 8 above)
Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Fixed.

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java (hunk 6 above)
Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Removed the variable

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java (hunk 4 above)
Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Modified the test case for the init() method.

Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Removed the assert and modified the test case for the init() method.

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java (hunk 3 above)
Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Fixed

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java (hunk 2 above)
Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Done

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java (hunk 1 above)
Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Done

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java (hunk 9 above)
Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/authcontext/DefaultClaimsRetrieverTest.java]: Yes. mocked realm service returns an object of User realm.

File: components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/bean/ScopeTest.java (hunk 10 above)
Comment (ChathurangiShyalika) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/bean/ScopeTest.java]: Defined two variables for the two constructors.

Comment (mefarazath) [reply components/org.wso2.carbon.identity.oauth/src/test/java/org/wso2/carbon/identity/oauth2/bean/ScopeTest.java]: @ChathurangiShyalika  I only see one at line 32 still. Am I missing something?

**Pull request merged**